### PR TITLE
feat(connector): [PowerTranz] Add cards 3ds support for PowerTranz connector

### PR DESCRIPTION
**Pull request by ArjunKarthik**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
3DS support for cards payment through powertranz connector

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context
To support 3ds cards through powertranz connector

## How did you test it?

Screen Shot 2023-07-16 at 11 29 43 PM

Screen Shot 2023-07-16 at 11 33 10 PM

Screen Shot 2023-07-16 at 11 29 58 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/connector/powertranz.rs [hunk 1]
```diff
@@ -230,6 +238,102 @@ impl ConnectorIntegration<api::Authorize, types::PaymentsAuthorizeData, types::P
...
+        let redirect_payload: powertranz::RedirectResponsePayload = req
+            .request
+            .redirect_response
+            .as_ref()
+            .and_then(|res| res.payload.to_owned())
+            .ok_or(errors::ConnectorError::MissingConnectorRedirectionPayload {
+                field_name: "request.redirect_response.payload",
+            })?
+            .parse_value("PowerTranz RedirectResponsePayload")
+            .change_context(errors::ConnectorError::ResponseDeserializationFailed)?;
```
Comment (jagan-jaya) [inline crates/router/src/connector/powertranz.rs:281]: move them to connector/utils
```suggestion
        let redirect_payload: powertranz::RedirectResponsePayload = req.to_redirect_payload()?;
```

**Review by jagan-jaya — CHANGES_REQUESTED** [review]

**Review by jagan-jaya — APPROVED** [review]

File: crates/router/src/connector/powertranz.rs [hunk 2]
```diff
@@ -230,6 +238,97 @@ impl ConnectorIntegration<api::Authorize, types::PaymentsAuthorizeData, types::P
...
+    fn get_request_body(
+        &self,
+        req: &types::PaymentsCompleteAuthorizeRouterData,
+    ) -> CustomResult<Option<types::RequestBody>, errors::ConnectorError> {
+        let redirect_payload: powertranz::RedirectResponsePayload = req
+            .request
+            .get_redirect_response_payload()?
+            .parse_value("PowerTranz RedirectResponsePayload")
+            .change_context(errors::ConnectorError::ResponseDeserializationFailed)?;
+        let spi_token = format!("\"{}\"", redirect_payload.spi_token);
```
Comment (SanchithHegde) [inline crates/router/src/connector/powertranz.rs:277]: You can use a raw string literal, you can take this up in a separate PR.

File: crates/router/src/connector/powertranz/transformers.rs [hunk 3]
```diff
@@ -55,6 +86,12 @@ pub struct PowertranzAddressDetails {
     phone_number: Option<Secret<String>>,
 }
 
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "PascalCase")]
+pub struct RedirectResponsePayload {
+    pub spi_token: String,
```
Comment (SanchithHegde) [inline crates/router/src/connector/powertranz/transformers.rs:92]: Should this be a secret?

File: crates/router/src/connector/powertranz/transformers.rs [hunk 4]
```diff
@@ -74,11 +117,43 @@ impl TryFrom<&types::PaymentsAuthorizeRouterData> for PowertranzPaymentsRequest
...
+        })
+    }
+}
+
+impl TryFrom<&types::PaymentsAuthorizeRouterData> for ExtendedData {
+    type Error = error_stack::Report<errors::ConnectorError>;
+    fn try_from(item: &types::PaymentsAuthorizeRouterData) -> Result<Self, Self::Error> {
+        Ok(Self {
+            three_d_secure: ThreeDSecure {
+                challenge_window_size: 5,
```
Comment (SanchithHegde) [inline crates/router/src/connector/powertranz/transformers.rs:135]: Why are we sending a hardcoded value for this field? Consider adding a comment in code if this is intentional.

**Review by SanchithHegde — APPROVED** [review]

File: crates/router/src/connector/powertranz/transformers.rs (hunk 3 above)
Comment (ArjunKarthik) [reply crates/router/src/connector/powertranz/transformers.rs:92]: This is not related to credential or PII info. we get this info in payment response also just a reference to retrieve 3ds payment

**Pull request merged**